src/dxc_helper.cc

```diff
@@ -0,0 +1,146 @@
+// limitations under the License.
+
+#include "src/dxc_helper.h"
+
+#include <algorithm>
+#include <sstream>
+
+#include "third_party/dxc/include/dxc/Support/Global.h"
+#include "third_party/dxc/include/dxc/Support/HLSLOptions.h"
+#include "third_party/dxc/include/dxc/Support/WinIncludes.h"
+#include "third_party/dxc/include/dxc/Support/dxcapi.use.h"
+#include "third_party/dxc/include/dxc/dxcapi.h"
+
+namespace amber {
+namespace dxchelper {
+namespace {
+
+const wchar_t* kDxcFlags[] = {
+    L"-spirv",  // SPIR-V compilation
+    L"-Vd"      // Disable validation.
+};
+const size_t kDxcFlagsCount = sizeof(kDxcFlags) / sizeof(const wchar_t*);
+
+// Converts an IDxcBlob into a vector of 32-bit unsigned integers which
+// is returned via the 'binaryWords' argument.
+void ConvertIDxcBlobToUint32(IDxcBlob* blob,
+                             std::vector<uint32_t>* binaryWords) {
+  size_t num32BitWords = (blob->GetBufferSize() + 3) / 4;
+  std::string binaryStr(static_cast<char*>(blob->GetBufferPointer()),
+                        blob->GetBufferSize());
+  binaryStr.resize(num32BitWords * 4, 0);
+  binaryWords->resize(num32BitWords, 0);
+  memcpy(binaryWords->data(), binaryStr.data(), binaryStr.size());
+}
+
+}  // namespace
+
+Result Compile(const std::string& src,
+               const std::string& entry,
+               const std::string& profile,
+               std::vector<uint32_t>* generated_binary) {
+  DxcInitThreadMalloc();
+
+  if (hlsl::options::initHlslOptTable()) {
+    DxcCleanupThreadMalloc();
+    return Result("DXC compile failure: initHlslOptTable");
+  }
+
+  IDxcLibrary* dxc_lib;
+  if (DxcCreateInstance(CLSID_DxcLibrary, IDxcLibrary::uuidof(),
+                        reinterpret_cast<void**>(&dxc_lib)) < 0) {
+    DxcCleanupThreadMalloc();
+    return Result("DXCCreateInstance for DXCLibrary failed");
+  }
+
+  IDxcBlobEncoding* source;
+  if (dxc_lib->CreateBlobWithEncodingOnHeapCopy(
+          src.data(), static_cast<uint32_t>(src.size()), CP_UTF8, &source) <
+      0) {
+    DxcCleanupThreadMalloc();
+    return Result("DXC compile failure: CreateBlobFromFile");
+  }
+
+  IDxcIncludeHandler* include_handler;
+  if (dxc_lib->CreateIncludeHandler(&include_handler) < 0) {
+    DxcCleanupThreadMalloc();
+    return Result("DXC compile failure: CreateIncludeHandler");
+  }
+
+  IDxcCompiler* compiler;
+  if (DxcCreateInstance(CLSID_DxcCompiler, IDxcCompiler::uuidof(),
+                        reinterpret_cast<void**>(&compiler)) < 0) {
+    DxcCleanupThreadMalloc();
+    return Result("DXCCreateInstance for DXCCompiler failed");
+  }
+
+  IDxcOperationResult* result;
+  std::wstring src_filename =
+      L"amber." + std::wstring(profile.begin(), profile.end());
+  if (compiler->Compile(source,               /* source text */
+                        src_filename.c_str(), /* original file source */
+                        std::wstring(entry.begin(), entry.end())
+                            .c_str(), /* entry point name */
```

Comment:
I had to stop for a while, worrying about the lifetime of the temporary.  But it dies at the end of the full expression, so that's late enough.  Phew.

Reply:
Heh, I didn't think about that at the time.